api/api.go

```diff
@@ -82,33 +98,85 @@ func (api *API) ListenAndServe(port int) error {
-			return
+		numRetriesLeft := DBTxnRetryCount
+		var body []byte
+		var err error
+		if req.Body != nil {
+			// Read the request's body and replace its Body io.ReadCloser with a
+			// new one based off the read data.
+			body, err = io.ReadAll(io.LimitReader(req.Body, LimitBodySizeLarge))
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to read body"), http.StatusBadRequest)
+				return
+			}
+			_ = req.Body.Close()
 		}
-		// Close session after the handler is done.
-		defer sess.EndSession(req.Context())
-
-		// Create session context.
-		sctx := mongo.NewSessionContext(req.Context(), sess)
 
-		// Get the special response writer.
-		mw, err := NewMongoWriter(w, sctx, api.staticLogger)
-		if err != nil {
-			api.WriteError(w, errors.AddContext(err, "failed to start a new transaction"), http.StatusInternalServerError)
-			return
+		// handleFn wraps a full execution of the handler, combined with a retry
+		// detection and counting. It also takes care of creating and cancelling
+		// Mongo sessions and transactions.
+		handleFn := func() (retry bool) {
+			req.Body = io.NopCloser(bytes.NewReader(body))
+			// Create a new db session
+			sess, err := api.staticDB.NewSession()
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to start a new mongo session"), http.StatusInternalServerError)
+				return
+			}
+			// Close session after the handler is done.
+			defer sess.EndSession(req.Context())
+			// Create session context.
+			sctx := mongo.NewSessionContext(req.Context(), sess)
+			// Get a special response writer which provide the necessary tools
+			// to retry requests on error.
+			mw, err := NewMongoWriter(w, sctx, api.staticLogger)
+			if err != nil {
+				api.WriteError(w, errors.AddContext(err, "failed to start a new transaction"), http.StatusInternalServerError)
+				return
+			}
+			// Create a new request with our session context.
+			req = req.WithContext(sctx)
+			// Forward the new response writer and request to the handler.
+			h(&mw, req, ps)
+
+			// If the call succeeded then we're done because both the status and
+			// the response content are already written to the response writer.
+			if mw.ErrorStatus() == 0 {
+				return false
+			}
+			// If the call failed with a WriteConflict error and we still have
+			// retries left, we'll retry it. Otherwise, we'll write the error to
+			// the response writer and finish the call.
+			if mw.FailedWithWriteConflict() && numRetriesLeft > 0 {
+				select {
+				case <-req.Context().Done():
+					// If the request context has expired we won't retry anymore.
+				default:
+					api.staticLogger.Tracef("Retrying call because of WriteConflict (%d out of %d). Request: %+v", numRetriesLeft, DBTxnRetryCount, req)
+					numRetriesLeft--
+					return true
+				}
+			}
+			// If the call failed with a non-WriteConflict error  or we ran out
```

Comment:
I think we definitely need an error log if we're out of retries to warn us that for whatever reason we can't get the transaction to pass even after retrying it. The 5 retries feels very arbitrary to me, especially since there's no exponential backoff mechanism or random x ms sleep.